[docs] Generate linting docs from source

jugglinmike opened the pull request:
When new rules have been added to WPT's "lint" tool, the corresponding
documentation has not always been updated [1] [2] [3]. The static list
of rules currently describes only 22 of the 53 available rules.
Automatically generating documentation from source code helps avoid this
state and the confusion it can cause contributors.

Rely on the previously-implemented source code structure [4] during
documentation generation to automatically create a listing of all
available linting rules.

Although the Sphinx documentation generator includes a built-in
extension for generating documentation from Python source code, the
output of that extension is designed to document Python primitives such
as functions and classes. Such a format is inappropriate for this case
because the users of the linting tool do not interact with the internals
in this way. Define a custom docutils directive to tailor the
documentation to the needs of its audience.

[1] https://github.com/web-platform-tests/wpt/issues/5299
[2] https://github.com/web-platform-tests/wpt/issues/10501
[3] https://github.com/web-platform-tests/wpt/issues/11479
[4] https://github.com/web-platform-tests/wpt/pull/16268

---

As noted in the commit message, this nearly doubles the number of documented
linting rules. Over time, we can further improve it by separating out explicit
"to fix" instructions (and maybe even require that new rules do the same).

jugglinmike committed:
[docs] Generate linting docs from source

When new rules have been added to WPT's "lint" tool, the corresponding
documentation has not always been updated [1] [2] [3]. The static list
of rules currently describes only 22 of the 53 available rules.
Automatically generating documentation from source code helps avoid this
state and the confusion it can cause contributors.

Rely on the previously-implemented source code structure [4] during
documentation generation to automatically create a listing of all
available linting rules.

Although the Sphinx documentation generator includes a built-in
extension for generating documentation from Python source code, the
output of that extension is designed to document Python primitives such
as functions and classes. Such a format is inappropriate for this case
because the users of the linting tool do not interact with the internals
in this way. Define a custom docutils directive to tailor the
documentation to the needs of its audience.

[1] https://github.com/web-platform-tests/wpt/issues/5299
[2] https://github.com/web-platform-tests/wpt/issues/10501
[3] https://github.com/web-platform-tests/wpt/issues/11479
[4] https://github.com/web-platform-tests/wpt/pull/16268

sideshowbarker reviewed (commented):
I suggest the existing descriptions from the `docs/writing-tests/lint-tool.md` file — rather than just being dropped — should be migrated over to the `tools/lint/rules.py` file, to replace the corresponding  existing embedded descriptions in that file. Rationale: The existing descriptions in the `docs/writing-tests/lint-tool.md` file in most cases seem more helpful than the corresponding  existing embedded descriptions in the `tools/lint/rules.py` file. That said, of course the descriptions should be evaluated case-by-case rather than just being replaced without reviewing them.

gsnedders commented on docs/wpt_lint_rules.py:
```diff
@@ -0,0 +1,78 @@
+
+            if rule["to_fix"]:
+                definition += nodes.strong(text="To fix:")
+                definition += WPTLintRules._parse_markdown(rule["to_fix"])
```
So we parse the `to_fix` as Markdown, but not the description?

jugglinmike replied:
Yup, that's a mistake. Good catch!

jugglinmike commented:
@sideshowbarker @gsnedders the descriptions are now parsed as Markdown, and the website's text has been preserved in all cases where it was more useful. Additionally, the list is lexicographically sorted now.

sideshowbarker reviewed (approved):
Beautiful

gsnedders commented on tools/lint/rules.py:
```diff
@@ -135,26 +139,42 @@ class NonexistentRef(Rule):
 class InvalidTimeout(Rule):
     name = "INVALID-TIMEOUT"
-    description = "Invalid timeout value %s"
+    description = (
```
Why aren't we just using a triple-quoted string here, as we did in earlier versions of the PR?

jugglinmike replied:
Multi-line strings are fine when rendering markup, but they aren't appropriate for the CLI. Since multi-line strings are easier to maintain, I experimented with normalizing the values at runtime using Python's standard library (specifically `textwrap.dedent` and `inspect.cleandoc`). That takes care of indentation, but the interstitial newlines are still a problem.

Rather than further increase complexity with custom formatting code, I opted for the uglier syntax.

sideshowbarker reviewed (approved):
Refinements all LGTM

jugglinmike commented:
@gsnedders do you support [my decision about multiline strings](https://github.com/web-platform-tests/wpt/pull/17189#discussion_r292092936)? If so, would you mind merging this?

gsnedders commented:
@jugglinmike Can you add a comment or something at the top justifying this?

I feel like this is something we're highly likely to regress in future, though. And probably add a parametrised test that goes through all the rules and checks the description/to_fix doesn't contain any newlines?

jugglinmike commented:
@gsnedders I suggested the requirement to author text without newlines as a compromise to limit complexity. In some ways, writing tests cuts against that goal. It's not that I'm against testing, but I'd rather test something that supports a more Pythonic contribution experience. That is, instead of complaining when folks naturally reach for multi-line strings, I'd rather try to support that pattern and test the logic which enables it.

Do you think it's better to treat this as an output sanitization problem? A utility function like this would be easy to test:

```python
def prep_for_command_line(text):
    return inspect.cleandoc(text).replace('\n', ' ')
```

gsnedders commented:
@jugglinmike we need more aggressive whitespace normalisation that that, don't we? or we'll just end up with lots of consecutive whitespace? (or does that work fine because it's interpreted as Markdown and whitespace is otherwise collapsed?)

jugglinmike commented:
I think the combination of [`inspect.cleandoc`](https://docs.python.org/2/library/inspect.html#inspect.cleandoc) and the removal of newlines should have us covered. Can you think of any specific cases where that wouldn't be enough?

```python
import inspect

strings = [
    '''a
       b
       c''',
    '''
    a
    b
    c''',
    '''
    a
    b
    c
    ''',
    '''a
    b
    c
    '''
]

for string in strings:
    print inspect.cleandoc(string).replace('\n', ' ')
```

    a b c
    a b c
    a b c
    a b c

gsnedders commented:
@jugglinmike I got side-tracked and forgot to check what `inspect.cleandoc` does 🙂

jugglinmike commented:
@gsnedders I now remember another wrinkle: the tests inspect the Python objects directly (as opposed to the CLI output). That means we can't clean these values as we write them out--they have to be stored in sanitized form.

I've added a function named `collapse` and called it for each multiline string value. This might be a little more ergonomic than string concatenation, but it still requires contributors to discover a local convention.

Another solution might be perform the `collapse` operation in a metaclass. That kind of trades "too explicit" for "too implicit" though.

Whatever we decide, future contributors will be made aware of extra whitespace without any new tests because the current tests assert the `description` value. In light of that (and these alternate approaches), do you have a preference for how we structure this code?

jugglinmike commented:
@gsnedders There was a strange failure in Taskcluster that seemed to be unrelated to the change

<details>
  <summary>logs</summary>

    E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python3-stdlib-extensions/python3-lib2to3_3.6.7-1~18.04_all.deb  404  Not Found [IP: 91.189.88.175 80]
    E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python3-stdlib-extensions/python3-distutils_3.6.7-1~18.04_all.deb  404  Not Found [IP: 91.189.88.175 80]
    E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python3.6/libpython3.6_3.6.7-1~18.04_amd64.deb  404  Not Found [IP: 91.189.88.175 80]
    E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python3.6/libpython3.6-dev_3.6.7-1~18.04_amd64.deb  404  Not Found [IP: 91.189.88.175 80]
    E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python-cryptography/python3-cryptography_2.1.4-1ubuntu1.2_amd64.deb  404  Not Found [IP: 91.189.88.175 80]
    E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/main/p/python3.6/python3.6-dev_3.6.7-1~18.04_amd64.deb  404  Not Found [IP: 91.189.88.175 80]
    E: Failed to fetch http://archive.ubuntu.com/ubuntu/pool/universe/p/python-pip/python3-pip_9.0.1-2.3~ubuntu1_all.deb  404  Not Found [IP: 91.189.88.175 80]
    E: Unable to fetch some archives, maybe run apt-get update or try with --fix-missing?
    [taskcluster 2019-09-05 17:55:57.623Z] === Task Finished ===
    [taskcluster 2019-09-05 17:55:57.701Z] Unsuccessful task run with exit code: 100 completed in 47.098 seconds

</details>

...except that it was consistent. I've rebased to try to address that ([the prior version of this branch is available here](https://github.com/web-platform-tests/wpt/compare/master...bocoup:docs-generate-lint-docs-orig-2)). Also fixed a problem flagged by the type system.

Would you mind taking another look at this?

jugglinmike commented:
@gsnedders gh-19952 improved the feedback we get for documentation patches. I've merged `master` into this branch so [we can see exactly what effect this pull request has on the generated content](https://github.com/web-platform-tests/wpt/runs/288437272). Does it look good to you?

jugglinmike commented:
I've rebased this as per [the recent advice to contributors](https://github.com/orgs/web-platform-tests/teams/reviewers/discussions/3). The prior version of this branch is available [here](https://github.com/web-platform-tests/wpt/compare/master...bocoup:docs-generate-lint-docs-orig-3).

@gsnedders Last call :)
